mettle/src/process.c

```diff
@@ -0,0 +1,376 @@
+#include <stdio.h>
+#include <errno.h>
+#include <stdarg.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <ctype.h>
+#include <fcntl.h>
+#include <pwd.h>
+#include <signal.h>
+#include <stdbool.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <wordexp.h>
+
+#include "log.h"
+#include "process.h"
+#include "buffer_queue.h"
+#include "uthash.h"
+
+struct process_queue {
+	struct ev_io w;
+	struct buffer_queue *queue;
+};
+
+struct process {
+	struct procmgr *mgr;
+
+	struct process_queue out, err;
+	process_read_cb_t out_cb, err_cb;
+	int in_fd;
+	int out_fd;
+	int err_fd;
+
+	struct ev_child cw;
+	process_exit_cb_t exit_cb;
+
+	void *cb_arg;
+
+	UT_hash_handle hh;
+	pid_t pid;
+};
+
+struct procmgr {
+	struct ev_loop *loop;
+	struct process *processes;
+};
+
+extern char **environ;
+
+pid_t process_get_pid(struct process *process)
+{
+	return process->pid;
+}
+
+static void free_process_queue(struct ev_loop *loop, struct process_queue *pipe)
+{
+	if (pipe->w.fd >= 0) {
+		close(pipe->w.fd);
+		ev_io_stop(loop, &pipe->w);
+	}
+
+	if (pipe->queue) {
+		buffer_queue_free(pipe->queue);
+		pipe->queue = NULL;
+	}
+}
+
+static void free_process(struct process *process)
+{
+	close(process->in_fd);
+	close(process->out_fd);
+	close(process->err_fd);
+	free_process_queue(process->mgr->loop, &process->out);
+	free_process_queue(process->mgr->loop, &process->err);
+	free(process);
+}
+
+static int switch_user(const char *user)
+{
+	struct passwd pwd, *result;
+	int bufsize = 8192;
+	void *buf = NULL;
+	int rc;
+	do {
+		buf = realloc(buf, bufsize);
+		if (buf == NULL) {
+			return -1;
+		}
+		rc = getpwnam_r(user, &pwd, buf, bufsize, &result);
+		bufsize += 8192;
+	} while (rc == ERANGE && bufsize < 1048576);
+
+	if (rc == 0) {
+		if (setuid(pwd.pw_uid))
+			exit(1);
```

Comment:
We should definitely print to stderr, and I would suggest `abort(3)` to exit the child, since any `atexit(3)` functions that are registered for the main loop (if we ever go down that path) would get called here, too. Also, I feel `abort(3)` is more semantically correct for these errors.